File: lib/parsers/hplParser.ts
import type { TuneData } from '@/types/tuneData'

// ─── Channel name aliases ─────────────────────────────────────────────────────
// HP Tuners VCM Scanner exports many variations — map them all to canonical keys.
// Aliases are matched after: trim, lowercase, strip surrounding quotes,
// collapse whitespace, remove trailing units in parentheses.

const CHANNEL_ALIASES: Record<string, string> = {
  // ── RPM ───────────────────────────────────────────────────────────────────
  'engine speed': 'RPM',
  'engine speed (rpm)': 'RPM',
  'enginespeed': 'RPM',
  rpm: 'RPM',
  'engine rpm': 'RPM',
  'rpm (rpm)': 'RPM',
  'eng speed': 'RPM',

  // ── TPS ───────────────────────────────────────────────────────────────────
  tps: 'TPS',
  'throttle position': 'TPS',
  'throttle position (tps)': 'TPS',
  'throttle position (%)': 'TPS',
  'throttle pos.': 'TPS',
  'throttle pos': 'TPS',
  'throttle (%)': 'TPS',
  'tps (%)': 'TPS',

  // ── MAP ───────────────────────────────────────────────────────────────────
  map: 'MAP',
  'manifold absolute pressure': 'MAP',
  'map (kpa)': 'MAP',
  'map (psi)': 'MAP',
  'map pressure': 'MAP',
  'baro/map': 'MAP',
  'manifold pressure': 'MAP',
  'map sensor': 'MAP',

  // ── MAF ───────────────────────────────────────────────────────────────────
  maf: 'MAF',
  'mass air flow': 'MAF',
  'maf (g/s)': 'MAF',
  'mass airflow': 'MAF',
  'air flow': 'MAF',
  'airflow (g/s)': 'MAF',
  'maf sensor': 'MAF',

  // ── STFT ──────────────────────────────────────────────────────────────────
  stft: 'STFT',
  'short term fuel trim': 'STFT',
  'short term fuel trim (%)': 'STFT',
  'stft (%)': 'STFT',
  'fuel trim short term': 'STFT',
  'st fuel trim': 'STFT',
  'stft b1': 'STFT',
  'short term fuel trim b1': 'STFT',
  'short term ft bank 1': 'STFT',
  'fuel trim st b1': 'STFT',
  'stft bank 1': 'STFT',
  'short term fuel trim bank 1': 'STFT',
  'st fuel trim (bank 1)': 'STFT',

  // ── LTFT ──────────────────────────────────────────────────────────────────
  ltft: 'LTFT',
  'long term fuel trim': 'LTFT',
  'long term fuel trim (%)': 'LTFT',
  'ltft (%)': 'LTFT',
  'fuel trim long term': 'LTFT',
  'lt fuel trim': 'LTFT',
  'ltft b1': 'LTFT',
  'long term fuel trim b1': 'LTFT',
  'long term ft bank 1': 'LTFT',
  'fuel trim lt b1': 'LTFT',
  'ltft bank 1': 'LTFT',
  'long term fuel trim bank 1': 'LTFT',
  'lt fuel trim (bank 1)': 'LTFT',

  // ── Lambda / O2 / AFR ─────────────────────────────────────────────────────
  o2: 'Lambda',
  lambda: 'Lambda',
  'o2 sensor': 'Lambda',
  'o2 (v)': 'Lambda',
  'wideband o2': 'Lambda',
  afr: 'Lambda',
  'air fuel ratio': 'Lambda',
  'equivalence ratio': 'Lambda',
  'wideband afr': 'Lambda',
  'wb afr': 'Lambda',
  'o2 b1 s1': 'Lambda',
  'o2 sensor b1 s1': 'Lambda',
  'fueling - afr commanded': 'Lambda',
  'commanded afr': 'Lambda',
  'target afr': 'Lambda',
  'lambda (λ)': 'Lambda',

  // ── Knock ─────────────────────────────────────────────────────────────────
  knock: 'KnockRetard',
  'knock retard': 'KnockRetard',
  'knock retard (deg)': 'KnockRetard',
  'spark knock retard': 'KnockRetard',
  'knock activity': 'KnockRetard',
  'ks retard': 'KnockRetard',
  'knock sensor retard': 'KnockRetard',
  'knock retard cyl avg': 'KnockRetard',
  'knock retard avg': 'KnockRetard',
  'total knock retard': 'KnockRetard',
  'ignition - knock retard': 'KnockRetard',
  'knock retard (°)': 'KnockRetard',

  // ── IAT ───────────────────────────────────────────────────────────────────
  iat: 'IAT',
  'intake air temp': 'IAT',
  'intake air temperature': 'IAT',
  'iat (°c)': 'IAT',
  'iat (°f)': 'IAT',
  'iat (c)': 'IAT',
  'iat (f)': 'IAT',
  'air temperature': 'IAT',
  'inlet air temp': 'IAT',
  'air temp': 'IAT',
  'intake temp': 'IAT',
  'air temperature (°c)': 'IAT',
  'air temperature (°f)': 'IAT',

  // ── Coolant ───────────────────────────────────────────────────────────────
  coolant: 'Coolant',
  'engine coolant temperature': 'Coolant',
  ect: 'Coolant',
  'coolant temp': 'Coolant',
  'coolant (°c)': 'Coolant',
  'coolant (°f)': 'Coolant',
  'coolant temp (°c)': 'Coolant',
  'coolant temp (°f)': 'Coolant',
  'engine coolant temp': 'Coolant',
  'water temp': 'Coolant',
  'coolant temperature': 'Coolant',

  // ── Boost ─────────────────────────────────────────────────────────────────
  boost: 'Boost',
  'boost pressure': 'Boost',
  'boost (psi)': 'Boost',
  'boost (kpa)': 'Boost',
  'turbo boost': 'Boost',
  'boost pressure (psi)': 'Boost',
  'boost pressure (kpa)': 'Boost',
  'turbo boost pressure': 'Boost',
}

const WANTED_CHANNELS = new Set([
  'RPM', 'TPS', 'MAP', 'MAF', 'STFT', 'LTFT',
  'Lambda', 'KnockRetard', 'IAT', 'Coolant', 'Boost',
])

// ─── Helpers ─────────────────────────────────────────────────────────────────

/** Strip surrounding quotes, trim, lowercase, collapse whitespace. */
function normaliseHeader(h: string): string {
  return h
    .trim()
    .replace(/^["']|["']$/g, '')   // strip surrounding " or '
    .trim()
    .toLowerCase()
    .replace(/\s+/g, ' ')
}

/** Normalise then also try stripping trailing unit block e.g. " (rpm)" → try without it too. */
function canonicalise(raw: string): string | null {
  const norm = normaliseHeader(raw)

  // Direct match
  if (CHANNEL_ALIASES[norm]) return CHANNEL_ALIASES[norm]
  if (WANTED_CHANNELS.has(norm.toUpperCase())) return norm.toUpperCase()

  // Try stripping trailing "(units)" block
  const withoutUnits = norm.replace(/\s*\([^)]*\)\s*$/, '').trim()
  if (withoutUnits !== norm && CHANNEL_ALIASES[withoutUnits]) return CHANNEL_ALIASES[withoutUnits]
  if (withoutUnits !== norm && WANTED_CHANNELS.has(withoutUnits.toUpperCase())) return withoutUnits.toUpperCase()

  return null
}

function looksNumeric(cell: string): boolean {
  const n = parseFloat(cell.trim().replace(/^["']|["']$/g, ''))
  return isFinite(n)
}

function isDataRow(cells: string[]): boolean {
  return cells.filter((c) => looksNumeric(c)).length >= Math.ceil(cells.length * 0.5)
}

function isUnitsRow(cells: string[]): boolean {
  // A units row has mostly non-numeric cells but no channel keyword hits
  const numericCount = cells.filter((c) => looksNumeric(c)).length
  if (numericCount > cells.length * 0.3) return false  // too many numbers → data row
  const channelHits = cells.filter((c) => canonicalise(c) !== null).length
  return channelHits === 0  // no channel names → probably units
}

function splitLine(line: string): string[] {
  // Prefer tab; fall back to comma
  if (line.includes('\t')) return line.split('\t')
  return line.split(',')
}

// ─── Main parser ─────────────────────────────────────────────────────────────

export async function parseHPL(file: File): Promise<TuneData> {
  const text = await file.text()
  const lines = text.split(/\r?\n/)

  let headerRowIndex = -1
  let columnHeaders: string[] = []

  // ── Step 1: find the column header row ──────────────────────────────────
  // Strategy: scan every row; prefer the first row with ≥1 recognised channel.
  // Fallback: first row where any non-blank row within the next 3 lines is all-numeric.
  for (let i = 0; i < Math.min(lines.length, 60); i++) {
    const line = lines[i]
    if (!line.trim()) continue

    const cells = splitLine(line)
    if (cells.length < 2) continue

    // Count recognisable channel names (threshold: 1 is enough)
    const matchCount = cells.filter((c) => canonicalise(c) !== null).length

    if (matchCount >= 1) {
      headerRowIndex = i
      columnHeaders = cells
      break
    }

    // Lookahead fallback: look up to 3 rows ahead for a numeric row
    for (let j = i + 1; j <= i + 3 && j < lines.length; j++) {
      const nextLine = lines[j]
      if (!nextLine.trim()) continue
      if (isDataRow(splitLine(nextLine)) && cells.length >= 3) {
        headerRowIndex = i
        columnHeaders = cells
        break
      }
    }
    if (headerRowIndex !== -1) break
  }

  if (headerRowIndex === -1 || columnHeaders.length === 0) {
    return buildEmptyTuneData('No channel header row found in .hpl file')
  }

  // ── Step 2: map column index → canonical channel name ───────────────────
  const colMap: Record<number, string> = {}
  columnHeaders.forEach((h, i) => {
    const canonical = canonicalise(h)
    if (canonical && WANTED_CHANNELS.has(canonical)) {
      colMap[i] = canonical
    }
  })

  const foundChannels = Object.values(colMap)

  if (foundChannels.length === 0) {
    // Diagnostics can still try to run on an empty datalog;
    // return the raw channel list so the user knows what was found
    const rawNames = columnHeaders
      .map((h) => normaliseHeader(h))
      .filter((h) => h.length > 0)
      .slice(0, 10)
      .join(', ')
    return buildEmptyTuneData(
      `No recognised channels found in .hpl file. ` +
      `Columns detected: ${rawNames || '(none)'}. ` +
      `Check that this is a VCM Scanner datalog and that channel names match expected HP Tuners format.`
    )
  }

  // ── Step 3: parse data rows, skipping units row if present ──────────────
  const dataRows: Record<string, number>[] = []
  let skippedUnitsRow = false

  for (let i = headerRowIndex + 1; i < lines.length; i++) {
    const line = lines[i]
    if (!line.trim()) continue

    const cells = splitLine(line)

    // Skip the units row (first non-blank row after header that has no channel hits and is not numeric)
    if (!skippedUnitsRow && isUnitsRow(cells)) {
      skippedUnitsRow = true
      continue
    }

    if (!isDataRow(cells)) continue

    const row: Record<string, number> = {}
    for (const [idxStr, channel] of Object.entries(colMap)) {
      const idx = Number(idxStr)
      const raw = cells[idx]?.trim().replace(/^["']|["']$/g, '')
      if (raw !== undefined && raw !== '') {
        const n = parseFloat(raw)
        if (isFinite(n)) row[channel] = n
      }
    }
    if (Object.keys(row).length > 0) dataRows.push(row)
  }

  // ── Step 4: derive flags ─────────────────────────────────────────────────
  const maxMAP   = dataRows.reduce((m, r) => Math.max(m, r.MAP   ?? 0), 0)
  const maxBoost = dataRows.reduce((m, r) => Math.max(m, r.Boost ?? 0), 0)
  const isForcedInduction = maxMAP > 110 || maxBoost > 2

  return {
    source: 'hpl',
    vehicle: {
      osVersion: null,
      ecuType: null,
      vin: null,
      platform: null,
    },
    tables: {},
    datalog: {
      channels: foundChannels,
      rows: dataRows,
    },
    rawFlags: {
      isForcedInduction,
      hasFlexFuel: false,
      hasBoostControl: isForcedInduction,
      detectedInjectorSize_cc: null,
    },
  }
}

function buildEmptyTuneData(warning: string): TuneData {
  return {
    source: 'hpl',
    vehicle: { osVersion: null, ecuType: null, vin: null, platform: null },
    tables: {},
    datalog: { channels: [], rows: [] },
    rawFlags: {
      isForcedInduction: false,
      hasFlexFuel: false,
      hasBoostControl: false,
      detectedInjectorSize_cc: null,
    },
    _parseWarning: warning,
  } as TuneData & { _parseWarning: string }
}

File: companion/parsers/hpl_parser.py
"""
VCM Scanner .hpl datalog parser.

Supports two formats:
  1. Text/CSV export — tab- or comma-separated rows with a header line.
  2. Binary HP Tuners native (.hpl "HPT " magic) — header is parsed for channel
     names, but the data section uses a proprietary encrypted format that cannot
     be decoded.  The parser returns a warning dict so the caller can prompt the
     user to export a text/CSV copy from VCM Scanner instead.
"""

from pathlib import Path
import struct

# ── Channel-name aliases (text format) ────────────────────────────────────────
CHANNEL_ALIASES: dict[str, str] = {
    # RPM
    "engine speed": "RPM", "engine speed (rpm)": "RPM",
    "enginespeed": "RPM", "rpm": "RPM", "engine rpm": "RPM",
    "rpm (rpm)": "RPM", "eng speed": "RPM",
    # TPS
    "tps": "TPS", "throttle position": "TPS",
    "throttle position (tps)": "TPS", "throttle position (%)": "TPS",
    "throttle pos.": "TPS", "throttle pos": "TPS",
    "throttle (%)": "TPS", "tps (%)": "TPS",
    # MAP
    "map": "MAP", "manifold absolute pressure": "MAP",
    "map (kpa)": "MAP", "map (psi)": "MAP", "map pressure": "MAP",
    "baro/map": "MAP", "manifold pressure": "MAP", "map sensor": "MAP",
    "map kpa": "MAP", "boost pressure": "MAP",
    # MAF
    "maf": "MAF", "mass air flow": "MAF",
    "maf (g/s)": "MAF", "mass airflow": "MAF",
    "massairflow": "MAF", "maf g/s": "MAF",
    "air flow": "MAF", "airflow (g/s)": "MAF", "maf sensor": "MAF",
    # STFT / LTFT  (map to combined keys; web app harmonises bank)
    "stft": "STFT", "short term fuel trim": "STFT",
    "short term fuel trim (%)": "STFT", "stft (%)": "STFT",
    "fuel trim short term": "STFT", "st fuel trim": "STFT",
    "stft b1": "STFT", "short term fuel trim b1": "STFT",
    "short term ft bank 1": "STFT", "fuel trim st b1": "STFT",
    "stft bank 1": "STFT", "short term fuel trim bank 1": "STFT",
    "st fuel trim (bank 1)": "STFT",
    "ltft": "LTFT", "long term fuel trim": "LTFT",
    "long term fuel trim (%)": "LTFT", "ltft (%)": "LTFT",
    "fuel trim long term": "LTFT", "lt fuel trim": "LTFT",
    "ltft b1": "LTFT", "long term fuel trim b1": "LTFT",
    "long term ft bank 1": "LTFT", "fuel trim lt b1": "LTFT",
    "ltft bank 1": "LTFT", "long term fuel trim bank 1": "LTFT",
    "lt fuel trim (bank 1)": "LTFT",
    # Lambda / O2 / AFR
    "o2": "Lambda", "lambda": "Lambda", "o2 sensor": "Lambda",
    "o2 (v)": "Lambda", "wideband o2": "Lambda", "afr": "Lambda",
    "air fuel ratio": "Lambda", "equivalence ratio": "Lambda",
    "wideband afr": "Lambda", "wb afr": "Lambda",
    "o2 b1 s1": "Lambda", "o2 sensor b1 s1": "Lambda",
    "fueling - afr commanded": "Lambda", "commanded afr": "Lambda",
    "target afr": "Lambda", "\u03bb": "Lambda",
    # Knock
    "knock": "KnockRetard", "knock retard": "KnockRetard",
    "knock retard (deg)": "KnockRetard", "spark knock retard": "KnockRetard",
    "knock activity": "KnockRetard", "ks retard": "KnockRetard",
    "knock sensor retard": "KnockRetard",
    "knock retard cyl avg": "KnockRetard", "knock retard avg": "KnockRetard",
    "total knock retard": "KnockRetard",
    "ignition - knock retard": "KnockRetard",
    # IAT
    "iat": "IAT", "intake air temp": "IAT",
    "intake air temperature": "IAT", "intakeairtemp": "IAT",
    "iat (\u00b0c)": "IAT", "iat (c)": "IAT",
    "iat (\u00b0f)": "IAT", "iat (f)": "IAT",
    "air temperature": "IAT", "inlet air temp": "IAT",
    "air temp": "IAT", "intake temp": "IAT",
    # CLT / Coolant
    "coolant": "CLT", "engine coolant temperature": "CLT",
    "ect": "CLT", "coolant temp": "CLT",
    "coolant (\u00b0c)": "CLT", "coolant (\u00b0f)": "CLT",
    "coolant temp (\u00b0c)": "CLT", "coolant temp (\u00b0f)": "CLT",
    "engine coolant temp": "CLT", "water temp": "CLT",
    "coolant temperature": "CLT",
    "enginecoolanttemp": "CLT", "ect \u00b0c": "CLT", "ect c": "CLT",
    # Boost (separate channel distinct from MAP)
    "boost": "Boost", "boost pressure": "Boost",
    "boost (psi)": "Boost", "boost (kpa)": "Boost",
    "turbo boost": "Boost", "boost pressure (psi)": "Boost",
    "boost pressure (kpa)": "Boost", "turbo boost pressure": "Boost",
}

RPM_LIKE  = {"rpm", "engine speed", "enginespeed", "engine speed (rpm)", "eng speed"}
MAP_LIKE  = {"map", "manifold absolute pressure", "map kpa", "boost pressure",
             "map (kpa)", "map (psi)", "manifold pressure"}
TPS_LIKE  = {"tps", "throttle position", "throttleposition", "throttle pos",
             "throttle position (tps)", "throttle position (%)"}

BINARY_HPL_MAGIC = b"HPT "   # HP Tuners native binary datalog signature

# ── Unit-string → canonical channel  (binary header channel info) ──────────────
_UNIT_TO_CANONICAL: dict[str, str] = {
    "rpm":    "RPM",
    "kpa":    "MAP",
    "g/s":    "MAF",
    "%":      "TPS",        # could also be STFT/LTFT — we label generically
    "v":      "Lambda",
    "\u03bb": "Lambda",
    "\u00b0": "IAT",        # bare '°' → temperature (IAT or CLT ambiguous)
    "\u00b0c": "CLT",
    "mpa":    "Boost",
}


# ─── Public entry point ────────────────────────────────────────────────────────

def parse_hpl(filepath: str) -> dict:
    """Return a TuneData-compatible dict for an HPL file.

    Binary HPL files cannot be decoded (proprietary encrypted format) but the
    channel names in the file header ARE readable.  In that case the function
    returns a dict with an empty datalog and a ``_parseWarning`` message so the
    caller can prompt the user to export a text/CSV copy from VCM Scanner.
    """
    with open(filepath, "rb") as f:
        raw_start = f.read(6)

    if raw_start[:4] == BINARY_HPL_MAGIC:
        return _handle_binary_hpl(filepath, raw_start)

    # ── Text/CSV format ────────────────────────────────────────────────────────
    with open(filepath, "r", encoding="utf-8", errors="ignore") as f:
        lines = f.readlines()

    sep = _detect_separator(lines)

    header_row = None
    data_start = 0
    for i, line in enumerate(lines):
        stripped = line.strip()
        if not stripped:
            continue
        cols = stripped.split(sep)
        lower_cols = [c.lower().strip().strip('"\'') for c in cols]
        hits = sum(
            1 for c in lower_cols
            if any(kw in c for kw in list(RPM_LIKE) + list(MAP_LIKE) + list(TPS_LIKE))
        )
        if hits > 0 or (len(cols) >= 3 and i < 20):
            header_row = cols
            data_start = i + 1
            break

    if not header_row:
        raise ValueError(
            "Could not find data headers in .hpl file. "
            "Ensure this is a VCM Scanner datalog exported as text/CSV."
        )

    normalised_headers = [_normalise_channel(h) for h in header_row]

    # Skip an optional units row (non-numeric row immediately after header)
    if data_start < len(lines):
        first_data = lines[data_start].strip()
        if first_data:
            test_cols = first_data.split(sep)
            num_count = sum(1 for c in test_cols if _is_numeric(c.strip().strip('"\'').strip()))
            if num_count < len(test_cols) * 0.3:
                data_start += 1

    rows = []
    for line in lines[data_start:]:
        stripped = line.strip()
        if not stripped:
            continue
        values = stripped.split(sep)
        row: dict[str, float | None] = {}
        for idx, col in enumerate(normalised_headers):
            if not col:
                continue
            try:
                raw = values[idx].strip().strip('"\'') if idx < len(values) else ""
                row[col] = float(raw) if raw else None
            except (ValueError, IndexError):
                row[col] = None
        if any(v is not None for v in row.values()):
            rows.append(row)

    is_fi = _detect_forced_induction(rows)

    return {
        "source": "companion_app",
        "vehicle": {
            "osVersion": None,
            "ecuType":   None,
            "vin":       None,
            "platform":  None,
        },
        "tables": {},
        "datalog": {
            "channels": normalised_headers,
            "rows":     rows,
        },
        "rawFlags": {
            "isForcedInduction": is_fi,
            "hasFlexFuel":       False,
            "hasBoostControl":   False,
            "detectedInjectorSize_cc": None,
        },
        "_parseWarning": None,
    }


# ── Binary HPL handling ────────────────────────────────────────────────────────

def _handle_binary_hpl(filepath: str, raw_start: bytes) -> dict:
    """Parse what we can from a binary HPL header and return a warning dict."""
    channels_found: list[str] = []
    n_channels = 0

    try:
        with open(filepath, "rb") as f:
            header = f.read(0x600)   # read first ~1.5 KB (covers all channel defs)

        # Version byte is at offset 5 (e.g. 0x09 for version 9)
        version = raw_start[5] if len(raw_start) > 5 else 0

        if version == 0x09:
            # Version 9: structured header with readable channel definitions.
            # Channel count at 0x27 (LE uint16).
            n_channels = struct.unpack_from("<H", header, 0x27)[0] if len(header) > 0x29 else 0
            channels_found = _extract_v9_channels(header, n_channels)

    except Exception:
        pass   # best-effort — fall through to warning

    ch_list = ""
    if channels_found:
        ch_list = (
            f"\n\nChannels detected in file header: {', '.join(channels_found)}"
            "\n(Channel data cannot be extracted from the binary format.)"
        )

    warning = (
        "Binary HP Tuners datalog detected.\n\n"
        "HP Tuners .hpl files saved directly by VCM Scanner use a proprietary "
        "encrypted format — the data cannot be read without HP Tuners software.\n\n"
        "To use this log:\n"
        "  1. Open the file in VCM Scanner\n"
        "  2. File \u2192 Export \u2192 Export to Text/CSV\n"
        "  3. Save as .csv or .hpl\n"
        "  4. Drop the exported file into this app"
        + ch_list
    )

    return {
        "source": "companion_app",
        "vehicle": {
            "osVersion": None,
            "ecuType":   None,
            "vin":       None,
            "platform":  None,
        },
        "tables": {},
        "datalog": {
            "channels": channels_found,
            "rows":     [],
        },
        "rawFlags": {
            "isForcedInduction": False,
            "hasFlexFuel":       False,
            "hasBoostControl":   False,
            "detectedInjectorSize_cc": None,
        },
        "_parseWarning": warning,
    }


def _extract_v9_channels(header: bytes, n_channels: int) -> list[str]:
    """Extract canonical channel names from a version-9 binary HPL header.

    Each channel definition contains (after the 8-byte scale=1.0 and 8-byte
    offset=0.0 fields): 1-byte padding, 1-byte unit-string length, the unit
    string (UTF-8), then 6 more bytes of per-channel metadata.
    """
    SCALE_1 = struct.pack("<d", 1.0)
    seen: dict[str, int] = {}
    found: list[str] = []
    pos = 0
    limit = min(len(header), 0x600)

    while len(found) < max(n_channels, 1) and pos < limit:
        idx = header.find(SCALE_1, pos, limit)
        if idx == -1:
            break
        after = idx + 16     # past scale (8) + offset (8)
        if after + 3 > len(header):
            break
        unit_len  = header[after + 1]
        unit_end  = after + 2 + unit_len
        if unit_end > len(header):
            break
        unit_raw = header[after + 2: unit_end]
        unit_str = unit_raw.decode("utf-8", errors="replace").strip().lower()
        canonical = _UNIT_TO_CANONICAL.get(unit_str)
        if canonical:
            count = seen.get(canonical, 0)
            label = canonical if count == 0 else f"{canonical}_{count + 1}"
            seen[canonical] = count + 1
            found.append(label)
        pos = idx + 1

    return found


# ── Helpers ────────────────────────────────────────────────────────────────────

def _detect_separator(lines: list[str]) -> str:
    tabs   = sum(line.count("\t") for line in lines[:20])
    commas = sum(line.count(",")  for line in lines[:20])
    return "\t" if tabs >= commas else ","


def _normalise_channel(name: str) -> str:
    key = name.strip().strip('"\'').lower().replace("  ", " ")
    return CHANNEL_ALIASES.get(key, name.strip().strip('"\''))


def _is_numeric(s: str) -> bool:
    try:
        float(s)
        return True
    except ValueError:
        return False


def _detect_forced_induction(rows: list[dict]) -> bool:
    for row in rows:
        map_val   = row.get("MAP")
        boost_val = row.get("Boost")
        if isinstance(map_val,   (int, float)) and map_val   > 110:
            return True
        if isinstance(boost_val, (int, float)) and boost_val > 2:
            return True
    return False
